modules/system-disaster-recovery/src/test/java/org/apache/ignite/internal/disaster/system/SystemDisasterRecoveryManagerImplTest.java

```diff
@@ -0,0 +1,490 @@
+    }
+
+    @Test
+    void resetClusterInitiatesRestartWhenMajorityOfCmgNodesRespondsWithOk() {
+        when(topologyService.allMembers()).thenReturn(List.of(thisNode, node2, node3, node4, node5));
+        prepareNodeStateForClusterReset();
+
+        respondSuccessfullyFrom(thisNode, node2);
+        respondWithExceptionFrom(node3, node4, node5);
+
+        CompletableFuture<Void> future = manager.resetCluster(List.of(thisNodeName, node2.name(), node3.name()));
+        assertThat(future, willCompleteSuccessfully());
+
+        verify(restarter).initiateRestart();
+    }
+
+    private void respondSuccessfullyFrom(ClusterNode... nodes) {
+        for (ClusterNode node : nodes) {
+            respondSuccessfullyFrom(node);
+        }
+    }
+
+    private void respondSuccessfullyFrom(ClusterNode node) {
+        when(messagingService.invoke(eq(node), any(), anyLong()))
+                .thenReturn(completedFuture(successResponseMessage));
+    }
+
+    private void respondWithExceptionFrom(ClusterNode... nodes) {
+        for (ClusterNode node : nodes) {
+            respondWithExceptionFrom(node);
+        }
+    }
+
+    private void respondWithExceptionFrom(ClusterNode node) {
+        when(messagingService.invoke(eq(node), any(), anyLong()))
+                .thenReturn(failedFuture(new TimeoutException()));
+    }
+
+    @Test
+    @DisplayName("resetCluster() fails and does not restart when majority of new CMG nodes do not respond")
+    void resetClusterFailsWhenNewCmgMajorityDoesNotRespond() {
+        when(topologyService.allMembers()).thenReturn(List.of(thisNode, node2, node3, node4, node5));
+        prepareNodeStateForClusterReset();
+
+        respondSuccessfullyFrom(thisNode, node4, node5);
+        respondWithExceptionFrom(node2, node3);
+
+        CompletableFuture<Void> future = manager.resetCluster(List.of(thisNodeName, node2.name(), node3.name()));
+        ClusterResetException ex = assertWillThrow(future, ClusterResetException.class, 10, SECONDS);
+        assertThat(ex.getMessage(), is("Did not get successful responses from new CMG majority, failing cluster reset."));
+
+        verify(restarter, never()).initiateRestart();
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void savesToVaultWhenGetsMessage(boolean fromSelf) throws Exception {
+        NetworkMessageHandler handler = extractMessageHandler();
+
+        ClusterNode conductor = fromSelf ? thisNode : node2;
+        handler.onReceived(resetClusterMessageOn2Nodes(conductor.name()), conductor, 0L);
+
+        assertTrue(waitForCondition(() -> vaultManager.get(new ByteArray(RESET_CLUSTER_MESSAGE_VAULT_KEY)) != null, 10_000));
+        VaultEntry entry = vaultManager.get(new ByteArray(RESET_CLUSTER_MESSAGE_VAULT_KEY));
+        assertThat(entry, is(notNullValue()));
+
+        ResetClusterMessage savedMessage = fromBytes(entry.value());
+
+        assertThat(savedMessage.cmgNodes(), containsInAnyOrder(thisNodeName, node2.name()));
```

Comment:
Maybe you could reuse `assertThatResetClusterMessageIsAsExpected` if you would keep it, because right now it's a copy-paste